tracer/src/Datadog.Trace/AppSec/Waf/Encoder.cs

```diff
@@ -302,5 +207,391 @@ private static StringBuilder FormatList(IEnumerable<object> objs, StringBuilder
+        {
+            var lstPointers = new List<IntPtr>();
+            var pool = Pool;
+            return new EncodeResult(lstPointers, pool, Encode(o, lstPointers, remainingDepth, key, applySafetyLimits, pool: pool));
+        }
+
+        public static unsafe DdwafObjectStruct Encode<TInstance>(TInstance? o, List<IntPtr> argToFree, int remainingDepth = WafConstants.MaxContainerDepth, string? key = null, bool applySafetyLimits = true, UnmanagedMemoryPool? pool = null)
+        {
+            pool ??= Pool;
+
+            DdwafObjectStruct ProcessKeyValuePairs<TKey, TValue>(IEnumerable<KeyValuePair<TKey, TValue>> enumerableDic, int count, delegate*<KeyValuePair<TKey, TValue>, string?> getKey, delegate*<KeyValuePair<TKey, TValue>, object?> getValue)
+                where TKey : notnull
+            {
+                var ddWafObjectMap = new DdwafObjectStruct { Type = DDWAF_OBJ_TYPE.DDWAF_OBJ_MAP };
+                if (!string.IsNullOrEmpty(key))
+                {
+                    ddWafObjectMap.ParameterName = ConvertToUtf8(key!, false);
+                    ddWafObjectMap.ParameterNameLength = (ulong)key!.Length;
+                }
+
+                if (applySafetyLimits)
+                {
+                    if (remainingDepth-- <= 0)
+                    {
+                        string GetItemsAsString()
+                        {
+                            var sb = StringBuilderCache.Acquire(StringBuilderCache.MaxBuilderSize);
+                            foreach (var x in enumerableDic)
+                            {
+                                sb.Append($"{getKey(x)}, {getValue(x)}, ");
+                            }
+
+                            if (sb.Length > 0)
+                            {
+                                sb.Remove(sb.Length - 2, 2);
+                            }
+
+                            return StringBuilderCache.GetStringAndRelease(sb);
+                        }
+
+                        Log.Warning("EncodeDictionary: object graph too deep, truncating nesting {Items}", GetItemsAsString());
+                        return ddWafObjectMap;
+                    }
+
+                    if (count > WafConstants.MaxContainerSize)
+                    {
+                        Log.Warning<int>("EncodeList: list too long, it will be truncated, MaxMapOrArrayLength {MaxMapOrArrayLength}", WafConstants.MaxContainerSize);
+                    }
+                }
+
+                var childrenCount = !applySafetyLimits || count < WafConstants.MaxContainerSize ? count : WafConstants.MaxContainerSize;
+                var childrenFromPool = ObjectStructSize * childrenCount < MaxBytesForMaxStringLength;
+                var childrenData = childrenFromPool ? pool.Rent() : Marshal.AllocCoTaskMem(ObjectStructSize * childrenCount);
+
+                if (enumerableDic is IDictionary)
+                {
+                    var typeKVP = typeof(KeyValuePair<TKey, TValue>);
```

Comment:
I'd cache the `typeof`s in static fields to avoid the Reflection slugginess

Reply:
`typeof` is already performant, in IL this is transformed into a `ldtoken` and then a `GetTypeFromHandle` that is really fast because is only a getter to the runtime handle field. https://source.dot.net/#System.Private.CoreLib/src/System/Type.CoreCLR.cs,97

So I don't think caching is required here.